File: eel-core/src/main/kotlin/io/eels/component/orc/OrcFns.kt
package io.eels.component.orc

import io.eels.schema.Field
import io.eels.schema.Schema
import org.apache.hadoop.fs.FileSystem
import org.apache.hadoop.fs.Path
import org.apache.hadoop.hive.ql.io.orc.OrcFile
import org.apache.hadoop.hive.ql.io.orc.RecordReader

object OrcFns {

  fun createOrcReader(path: Path, fs: FileSystem): RecordReader = OrcFile.createReader(fs, path).rows()

  fun orcSchemaFromReader(reader: RecordReader): Schema {
    val fields = reader.next(null) as List<Any?>
    val columns = fields.map { Field(it.toString()) }
    return Schema(columns)
  }
}

File: eel-core/src/main/scala/io/eels/component/orc/OrcFns.scala
package io.eels.component.orc

import io.eels.schema.Schema
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.Path
import org.apache.orc.OrcFile.ReaderOptions
import org.apache.orc.{OrcFile, TypeDescription}

import scala.collection.JavaConverters._

object OrcFns {

  def writeSchema(schema: Schema): TypeDescription = {
    val orcSchema = TypeDescription.createStruct()
    schema.fields.foreach { field =>
      orcSchema.addField(field.name, TypeDescription.createString())
    }
    orcSchema
  }

  def readSchema(schema: TypeDescription): Schema = Schema.fromFieldNames(schema.getFieldNames.asScala)

  def readSchema(path: Path)(implicit conf: Configuration): Schema = {
    val reader = OrcFile.createReader(path, new ReaderOptions(conf).maxLength(1))
    val schema = reader.getSchema()
    readSchema(schema)
  }
}